packages/sources/twosigma/src/endpoint/price.ts

```diff
@@ -0,0 +1,182 @@
+import { ProviderResult, makeLogger } from '@chainlink/external-adapter-framework/util'
+
+import { customSettings } from '../config'
+
+// Schema of message sent to Two Sigma to start streaming symbols
+export type WebSocketRequest = {
+  api_key: string
+  symbols: string[] // e.g. ["AAPL/USD"]
+}
+
+// Schema of message sent from Two Sigma containing symbol prices
+export type WebSocketMessage = {
+  timestamp: number // The UTC timestamp in seconds. e.g. 1666705053.713266
+  symbol_price_dict: Record<
+    string, // The ticker symbol. e.g. AAPL/USD
+    SymbolPriceData
+  >
+}
+
+export type SymbolPriceData = {
+  quote_currency: string // Quote currency. e.g. USD
+  session_status_flag:
+    | 'premarket' // Pre-market trading during pre-market hours (before 9.30am)
+    | 'open' // Market is open and trading during market hours
+    | 'postmarket' // After-hours trading during post-market hours
+    | 'closed' // Market is closed
+  asset_status_flag:
+    | 'active' // TSS is trading this stock actively
+    | 'inactive' // TSS has not traded this stock recently
+  confidence_interval: number // Confidence interval in quote currency units. e.g. 0.16416514635149188
+  price: number // Price in quote currency units. e.g. 379.64
+}
+
+export type WebSocketEndpointTypes = {
+  Request: {
+    // i.e. { base, quote }
+    // base is the symbol to query, e.g. AAPL
+    // quote is the currency to convert to, e.g. USD
+    Params: PriceEndpointParams
+  }
+  Response: {
+    Data: {
+      result: number
+    }
+    Result: number
+  }
+  CustomSettings: typeof customSettings
+  Provider: {
+    WsMessage: WebSocketMessage
+  }
+}
+
+const logger = makeLogger('TwoSigmaPriceWebsocketEndpoint')
+
+export class TwoSigmaWebsocketTransport extends WebSocketTransport<WebSocketEndpointTypes> {
+  // The API works by first sending a message to the server containing the API key
+  // and a list of ticker symbols to subscribe to, after which the server begins
+  // streaming price updates. However, in order to subscribe to a different set of
+  // symbols, the client would need to do so on a new websocket connection as the
+  // server seems to ignore all messages after the first.
+
+  override async streamHandler(
+    context: EndpointContext<WebSocketEndpointTypes>,
+    subscriptions: SubscriptionDeltas<WebSocketEndpointTypes['Request']['Params']>,
+  ): Promise<void> {
+    if (this.wsConnection && (subscriptions.new.length || subscriptions.stale.length)) {
+      logger.debug(
+        `closing WS connection for new subscriptions: ${JSON.stringify(subscriptions.desired)}`,
+      )
+
+      // 'close' is an async method that doesn't return a promise. Rather than
+      // polling and waiting for the connection to close, we can simply set the
+      // variable to null and rely on the parent method to create a new connection.
+      this.wsConnection.close()
+      this.wsConnection = null
```

Comment:
This is potentially dangerous if the connections hang indefinitely and we end up continually creating new ones, but we have metrics that should surface this situation if it were to occur

Reply:
I can also add an async check to see if the conn hasn't closed in 1s and emit a log if so.